### PR TITLE
fix(@aws-amplify/datastore): fix consecutive updates

### PR DESCRIPTION
**Pull request by iartemiev**
_Issue #, if available:_ https://github.com/aws-amplify/amplify-js/issues/6363

```graphql
type Todo @model {
  id: ID!
  name: String!
}
```

Currently, if we run the following code in DataStore:

```js
async function changeTodoName() {
  const [todo] = await DataStore.query(Todo); // { name: A,  _version: 1 }

  for (let letter of ['B', 'C', 'D']) {
    await DataStore.save(
      Todo.copyOf(todo, (updated) => {
        updated.name = letter;
      })
    );
  }
}
```

...the name of our Todo will be `'B'` after all AppSync requests are resolved, i.e., latter updates get ignored.

The expected behavior is for `name` to be `'D'`.

Here's a sequence diagram detailing the current behavior (abstracted/simplified):
consecutive writes seq diagram

It happens on both fast and slow networks.

**Fast (~500 Mbps):**
fast-network

**Slow (10 Kbps):**
slow-network

### Proposal
Reconcile the version of update mutations in the mutation queue when an AppSync mutation for the same model ID resolves.

Diagram of updated behavior:
consecutive writes seq diagram postpr

Here's the same example running with the code in this PR:
**Fast (~500 Mbps):**
fast-network-pr

**Slow (10 Kbps):**
slow-network-pr

**Note: the `_version` in the response will only be synced to related items in the queue when the AppSync mutation response contains the same item that was sent in the request, i.e., if the original mutation was sent by the same client.
In the event of a handled conflict, where the response payload is different from the request payload, we do not sync the `_version` and subsequent writes will effectively be discarded.**

Sequence diagram for multi-client behavior:
DataStore Consecutive Updates - Multi #2

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nubpro** [comment]
Been wondering about this, how does iOS and Android library counterpart deal with such issue?
Do they share the same implementation?

tagging @jamesonwilliams perhaps he can share some insides from the Android side.

**Comment by iartemiev** [comment]
@nubpro they most likely have the same issue on iOS and Android. We're going to sync up on this across teams after the holidays.

**Comment by kylanry** [comment]
Hi guys! Does anybody know when this fix will be added to the new Amplify release? I am using the iOS Amplify SDK and also got the same issue, it is blocker for us. Will this fix also applied for iOs version?

**Comment by iartemiev** [comment]
> Hi guys! Does anybody know when this fix will be added to the new Amplify release? I am using the iOS Amplify SDK and also got the same issue. It is blocker for us. Will this fix also applied for iOS version?

@kylanry, yes, this fix will be applied across the Amplify libraries (JS, iOS, Android, Flutter). We're making some final modifications to the solution and are planning on releasing it soon.

File: packages/datastore/src/sync/processors/mutation.ts [hunk 1]
```diff
@@ -151,14 +151,22 @@ class MutationProcessor {
...
 			const record = result.data[opName];
-			await this.outbox.dequeue(this.storage);
-
-			const hasMore = (await this.outbox.peek(this.storage)) !== undefined;
+			let hasMore = false;
+
+			await this.storage.runExclusive(async storage => {
+				// using runExclusive to prevent possible race condition
+				// when another record gets enqueued between dequeue and peek
+				console.log('dequeue record', record);
```
Comment (nubpro) [inline packages/datastore/src/sync/processors/mutation.ts:166]: Might want to remove this `console.log`?

**Comment by iartemiev** [comment]
@manueliglesias @amhinson - this is ready for review now 😸 

`outbox.test.ts` is very hairy because I had to do a ton of manual dependency injection to test the outbox functionality without the Sync Engine running.

**Review by amhinson — APPROVED** [review]
Looks good to me 👍  Great job on the tests!

File: scripts/build.js [hunk 2]
```diff
@@ -157,7 +157,13 @@ async function buildES5(typeScriptCompiler) {
 	let compilerOptions = {
 		esModuleInterop: true,
 		noImplicitAny: false,
-		lib: ['dom', 'es2017', 'esnext.asynciterable', 'es2018.asyncgenerator'],
+		lib: [
+			'dom',
+			'es2017',
+			'esnext.asynciterable',
+			'es2018.asyncgenerator',
+			'es2019.object',
```
Comment (amhinson) [inline scripts/build.js:165]: Just curious, why was this needed?

Comment (iartemiev) [reply scripts/build.js:165]: I added `es2019.object` in order to use `Object.fromEntries` [here](https://github.com/aws-amplify/amplify-js/pull/7354/files#diff-89f8afda2fb41bbdaea63170a71ab23374df7b81bef86a6f7cf1eb9af0baea78R456-R457). I have actually found myself needing it before, but had avoided it since we didn't have that in our build target. Figured now's as good of a time as any to add it.

**Comment by nubpro** [comment]
Just want to drop by to say that the included diagrams are supppper duper helpful, I wish there's more......could you guys spill more beans on the rest of datastore ins and outs?

I was gonna try to understand the solution u guys came out with and honestly, I wasn't able to totally comprehend them.
Mad props to @iartemiev and the team for working on this!!
